contracts/service-provider-directory/src/contract/execute.rs

```diff
@@ -0,0 +1,133 @@
+    let current_entries = query::query_owner(deps, owner.clone())?;
+    if current_entries.services.len() < MAX_NUMBER_OF_PROVIDERS_PER_OWNER as usize {
+        Ok(())
+    } else {
+        Err(ContractError::ReachedMaxProvidersForAdmin {
+            max_providers: MAX_NUMBER_OF_PROVIDERS_PER_OWNER,
+            owner,
+        })
+    }
+}
+
+fn ensure_max_aliases_per_nym_address(deps: Deps, nym_address: NymAddress) -> Result<()> {
+    let current_entries = query::query_nym_address(deps, nym_address.clone())?;
+    if current_entries.services.len() < MAX_NUMBER_OF_ALIASES_FOR_NYM_ADDRESS as usize {
+        Ok(())
+    } else {
+        Err(ContractError::ReachedMaxAliasesForNymAddress {
+            max_aliases: MAX_NUMBER_OF_ALIASES_FOR_NYM_ADDRESS,
+            nym_address,
+        })
+    }
+}
+
+fn ensure_service_exists(deps: Deps, service_id: ServiceId) -> Result<()> {
+    if state::services::has_service(deps.storage, service_id) {
+        Ok(())
+    } else {
+        Err(ContractError::NotFound { service_id })
+    }
+}
+
+fn ensure_sender_authorized(info: MessageInfo, service: &Service) -> Result<()> {
+    if info.sender == service.owner {
+        Ok(())
+    } else {
+        Err(ContractError::Unauthorized {
+            sender: info.sender,
+        })
+    }
+}
+
+/// Announce a new service. It will be assigned a new service provider id.
+pub fn announce(
+    deps: DepsMut,
+    env: Env,
+    info: MessageInfo,
+    nym_address: NymAddress,
+    service_type: ServiceType,
+) -> Result<Response> {
+    ensure_max_services_per_owner(deps.as_ref(), info.sender.clone())?;
+    ensure_max_aliases_per_nym_address(deps.as_ref(), nym_address.clone())?;
+
+    let deposit_required = state::deposit_required(deps.storage)?;
+    let denom = deposit_required.denom.clone();
+    let will_deposit = cw_utils::must_pay(&info, &denom)
+        .map_err(|err| ContractError::DepositRequired { source: err })?;
+    ensure_correct_deposit(will_deposit, deposit_required.amount)?;
+
+    let new_service = Service {
+        nym_address,
+        service_type,
+        owner: info.sender,
+        block_height: env.block.height,
+        deposit: Coin::new(will_deposit.u128(), denom),
+    };
+    let service_id = state::services::save(deps.storage, &new_service)?;
+    Ok(Response::new()
+        .add_attribute("action", "announce")
+        .add_attribute("service_id", service_id.to_string())
+        .add_attribute("service_type", service_type.to_string()))
+}
+
+/// Delete an exsisting service.
+pub fn delete(deps: DepsMut, info: MessageInfo, service_id: ServiceId) -> Result<Response> {
+    ensure_service_exists(deps.as_ref(), service_id)?;
+    let service_to_delete = state::services::load_id(deps.storage, service_id)?;
+    ensure_sender_authorized(info, &service_to_delete)?;
+
+    // TODO: should this be reduced to take transaction costs into account? So that the contract
```

Comment:
nope. contract doesn't pay any transaction fees. Those are all paid by the caller.